### ISSUE TITLE
[Draft] ELP Staking Pools

### ISSUE THREAD

**Issue by dmvt** [ElasticSwap/token#5]
# Description

We need a set of contracts to allow for a new form of staking which blends the ve and x staking models. 

The contract allows you to lock $TIC and earn rewards on a per second basis, in the same way the current StakingPools contract works. The Stake and Pools libraries can be used as is or adapted. The contract will also have the ability to process [ELP tokens minted as fees from ElasticSwap exchange pools](https://github.com/ElasticSwap/elasticswap/blob/develop/src/contracts/Exchange.sol#L256). ELP will be unwrapped and the underlying assets sold for a specified fee pair token. The fee pair token will then be paired with minted TIC up to the available emissions and used to mint ELP tokens. These tokens will be available as rewards to the stakers proportionally to their deposits.

# Draft Documentation

## ELP Staking Pools

The ELP Staking Pools work like a hybrid between a `ve` staking contract and a `x` staking contract. They are designed to accrue $TIC, but disburse it in the form of an ElasticSwap LP ($ELP) position as the platform accrues fees.

When you stake a token in an ELP Staking Pool, it is not locked. Rewards start to accrue instantly, but in an unrealized form. This means that you are entitled to a reward distribution of that amount, but only once enough fees have been collected to generate the actual $ELP tokens. ElasticSwap charges a 50 bps fee whenever a liquidity provider exits one of our exchange pools. This fee is received by the ELP Staking Pool contract and converted into $ELP rewards. The rewarded $ELP represents a liquidity position in the $TIC/USDC.e pool. Once you receive can feel free to restake these tokens, exit for underlying tokens, or anything in between.

## Why?

This yield farming approach is designed to align the incentives of staking liquidity providers and the protocol itself. By tying staking reward distribution to fee generation, we ensure that the community is rewarded for promoting the platform, not just for buying low and selling high.

### Why not ve?

With the `ve` model, stakers are rewarded based on how long they are willing to lock their tokens. This prevents stakers from withdrawing and liquidating their original deposit in an emergency situation. It also encourages the forming of cartels who gather the token to be staked, stake it receiving the veToken in return, and the wrap that veToken to make it liquid again, defeating the entire purpose of the locking period.

### What kind of APR can I expect?

We have come to the conclusion that APR is the wrong metric to look at. Instead, we display unrealized rewards and claimable rewards. Unrealized rewards are $TIC waiting for protocol fees to be generated. These fees are realized when an LP in any ElasticSwap exchange exits their position. Once a day, the collected fees are paired with newly minted $TIC, minting new claimable $ELP rewards.

### Can I withdraw my $TIC before the protocol generates fees?

Yes. However, you will forfeit any unrealized rewards.

### What happens to rewards I forfeit?

They are distributed to the remaining stakers purportionally.

### How long will I have to stake to receive all of my rewards?

That depends on how many fees the protocol is generating. The more fees, the faster you will be able to claim your rewards.

# Interface

The following interface should be used.

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity 0.8.4;

import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";

/// @title IELPStakingPools
/// @dev A contract interface which allows users to stake tokens in exchange for ELP rewards.
///
/// This contract was inspired by Chef Nomi's 'MasterChef' contract which can be found in this
/// repository: https://github.com/sushiswap/sushiswap.
interface IELPStakingPools is ReentrancyGuard {
  /// @dev Sets the governance.
  ///
  /// This function can only called by the current governance.
  ///
  /// emit PendingGovernanceUpdated(_pendingGovernance);
  ///
  /// @param _pendingGovernance the new pending governance.
  function setPendingGovernance(address _pendingGovernance) external;

  /// @dev Accepts the governance role.
  ///
  /// This function can only called by the pending governance.
  ///
  /// emit GovernanceUpdated(_pendingGovernance);
  function acceptGovernance() external;

  /// @dev Sets the potential distribution reward rate.
  ///
  /// This function can only called by the current governance.
  ///
  /// This will update all of the pools.
  ///
  /// @param _rewardRate The number of tokens to accrue per second.
  function setRewardRate(uint256 _rewardRate) external;

  /// @dev Creates a new pool.
  ///
  /// The created pool will need to have its reward weight initialized before it begins generating rewards.
  ///
  /// This function can only called by the current governance.
  ///
  /// emit PoolCreated(_poolId, _token);
  ///
  /// @param _token The token the pool will accept for staking.
  ///
  /// @return the identifier for the newly created pool.
  function createPool(IERC20 _token) external returns (uint256);

  /// @dev Adds an ELP token to the fee list. A token must be in this list to be unwound by the
  /// processFees function.
  ///
  /// This function can only called by the current governance.
  ///
  /// emit FeeTokenAdded(_token);
  ///
  /// @param _token the address of the token to be added.
  /// @param _liquidateToken the address of either the base or quote token.
  ///
  /// @return true
  function addFeeToken(_token, _liquidateToken) external returns (bool);

  /// @dev Sets the reward weights of all of the pools.
  ///
  /// @param _rewardWeights The reward weights of all of the pools.
  function setRewardWeights(uint256[] calldata _rewardWeights) external;

  /// @dev Stakes tokens into a pool.
  ///
  /// @param _poolId        the pool to deposit tokens into.
  /// @param _depositAmount the amount of tokens to deposit.
  function deposit(uint256 _poolId, uint256 _depositAmount) external;

  /// @dev Exits the LP position of any ELP fee tokens available and sells _liquidateToken back to
  /// the ELP's exchange contract. Any reward ELP that can be minted is minted.
  ///
  /// emit FeesProcessed();
  ///
  /// @return true
  function processFees() external returns (bool);

  /// @dev Converts one token to another using the specified ELP pool. Any reward ELP that can be
  /// minted is minted.
  ///
  /// This function can only called by the current governance.
  ///
  /// emit FeesProcessed();
  ///
  /// @param _token the address of the token to be liquidated.
  /// @param _exchange the ElasticSwap exchange to use for the liquidation.
  ///
  /// @return true
  function convertFeeToken(address _token, address _exchange) external returns (bool);

  /// @dev Rescues a token from the pool, transferring it to the governance address.
  ///
  /// This function can only called by the current governance.
  ///
  /// @param _token The token to rescue
  ///
  /// @return true
  function rescue(_token) external returns (bool);

  /// @dev Withdraws staked tokens from a pool. Redistributes any owed rewards to other stakers
  /// proportionally to their current stake. The rewards redistributed are relative to the
  /// amount withdrawn, so if only half of the staked tokens are withdrawn, half of the rewards
  /// are redistributed. No rewards are disbursed by this function.
  ///
  /// @param _poolId          The pool to withdraw staked tokens from.
  /// @param _withdrawAmount  The number of tokens to withdraw.
  function withdraw(uint256 _poolId, uint256 _withdrawAmount) external;

  /// @dev Claims all rewarded ELP tokens from a pool. Does not claim unpaired rewards.
  ///
  /// @param _poolId The pool to claim rewards from.
  ///
  /// @notice use this function to claim the ELP tokens from a corresponding pool by ID.
  function claim(uint256 _poolId) external;

  /// @dev Claims all ELP rewards from a pool and then withdraws all staked tokens. Any non-paired
  /// rewards are redistributed to the remaining stakers proportionally.
  ///
  /// @param _poolId the pool to exit from.
  function exit(uint256 _poolId) external;

  /// @dev Gets the rate at which tokens are minted to stakers for all pools.
  ///
  /// @return the reward rate.
  function rewardRate() external view returns (uint256);

  /// @dev Gets the total reward weight between all the pools.
  ///
  /// @return the total reward weight.
  function totalRewardWeight() external view returns (uint256);

  /// @dev Gets the number of pools that exist.
  ///
  /// @return the pool count.
  function poolCount() external view returns (uint256);

  /// @dev Gets the token a pool accepts.
  ///
  /// @param _poolId the identifier of the pool.
  ///
  /// @return the token.
  function getPoolToken(uint256 _poolId) external view returns (IERC20);

  /// @dev Gets the total amount of funds staked in a pool.
  ///
  /// @param _poolId the identifier of the pool.
  ///
  /// @return the total amount of staked or deposited tokens.
  function getPoolTotalDeposited(uint256 _poolId) external view returns (uint256);

  /// @dev Gets the reward weight of a pool which determines how much of the total rewards it receives per second.
  ///
  /// @param _poolId the identifier of the pool.
  ///
  /// @return the pool reward weight.
  function getPoolRewardWeight(uint256 _poolId) external view returns (uint256);

  /// @dev Gets the amount of tokens per second being accrued to stakers for a pool.
  ///
  /// @param _poolId the identifier of the pool.
  ///
  /// @return the pool reward rate.
  function getPoolRewardRate(uint256 _poolId) external view returns (uint256);

  /// @dev Gets the number of tokens a user has staked into a pool.
  ///
  /// @param _account The account to query.
  /// @param _poolId  the identifier of the pool.
  ///
  /// @return the amount of deposited tokens.
  function getStakeTotalDeposited(address _account, uint256 _poolId) external view returns (uint256);

  /// @dev Gets the number of unclaimed ELP tokens a user can claim from a pool.
  ///
  /// @param _account The account to get the unclaimed balance of.
  /// @param _poolId  The pool to check for unclaimed rewards.
  ///
  /// @return the amount of unclaimed ELP tokens a user has in a pool.
  function getStakeTotalUnclaimed(address _account, uint256 _poolId) external view returns (uint256);

  /// @dev Gets the number of rewards tokens a user can claim from a pool once they are paired.
  ///
  /// @param _account The account to get the unclaimed balance of.
  /// @param _poolId  The pool to check for unclaimed rewards.
  ///
  /// @return the amount of unrealized reward tokens a user has in a pool.
  function getStakeTotalUnrealized(address _account, uint256 _poolId) external view returns (uint256);
}
```

# Definition of done

- The above interface is fully implemented and behaves according to the function comments.
- There is 100% test coverage proving the above.
- There is a deploy script for the new contract.

**Comment by 0xean** [comment]
Working through the technical details of this:

When we get the fee ELP from pair X vs Y, we can remove our liquidity and now we have X and Y.  Turning  X and Y into TIC and USDC entirely on-chain isn't easy.  We don't know for sure that we (or sushi for that matter) have a viable route to get from X ->  USDC. 

I think for version 1 doing this in a DAO treasury trusted manner may make more sense.  The ELP tokens generated go to the DAO address.  The DAO is responsible for 1) turning these into USDC 2) depositing USDC into our staking contract which creates the TIC<>USD ELP that we reward to stakers. 3) providing monthly accounting for transparency (all of this ofc is also on-chain).  

This also removes the "oracle" problem of trusting the on-chain swap price to not be manipulated without additional human checks.

Essentially we end up with a function like

```
  function rewardStakers(uint256 amount) external {
    // 1. transfers USDC amount from caller to this contract
    // 2. Generates TIC<>USDC ELP token
    // 3. Updates staking pools with claimable amounts of ELP
    // 4. emits event
  }
```

Tokens that don't have any way to swap USDC can be held in the DAO treasury until a better option becomes available.


A few other points.  TIC<>USDC that is created will be claimable in a pro-rata fashion against a users unclaimed reward balance.  For example, I have 40 TIC unclaimed, and there is 4000 unclaimed TIC in the system.  I would be able to claim 40/4000 of any available TIC<>USDC ELP available.

Finally, I would like to propose a different mechanism when a user un-stakes and forfeits their unrealized TIC.  I believe this unclaimed TIC should go back to the DAO. The DAO would therefore accumulate TIC<>USDC LP in the future, guaranteeing protocol owned liquidity benefiting all token holders. 


Working on these contracts now!

**Comment by dmvt** [comment]
> I think for version 1 doing this in a DAO treasury trusted manner may make more sense. The ELP tokens generated go to the DAO address. The DAO is responsible for 1) turning these into USDC 2) depositing USDC into our staking contract which creates the TIC<>USD ELP that we reward to stakers. 3) providing monthly accounting for transparency (all of this ofc is also on-chain).

Makes total sense to me. Let's keep the development fast and simple.

> Finally, I would like to propose a different mechanism when a user un-stakes and forfeits their unrealized TIC. I believe this unclaimed TIC should go back to the DAO. The DAO would therefore accumulate TIC<>USDC LP in the future, guaranteeing protocol owned liquidity benefiting all token holders.

I think this is a great add.

**Comment by 0xean** [comment]
Adding a bit more info to the problem.

Example:

At time t=0, we have 10 stakers who have 100 unclaimed TIC, for sake of the example, they are all of equal stake size. 

At time t=1, we generate $100 USDC in fees and assuming TIC = 10$, we mint LP tokens consuming 10 TIC. Each staker should be able to claim ~1 TIC worth of LP

At time t=2, whale staker 1 comes along, and stakes 500 TIC. 

At time t=3, whale staker 1 now has 50 unclaimed TIC, but we need to make sure he cannot claim ANY of the LP since he was not staked from t=0 to t=1

Computing this on-chain becomes expensive, off chain it's very simple.

**Comment by 0xean** [comment]
closed in #6
